# Worksheet  11 #
Midterm 1 with emphasis on functions

In [1]:
%matplotlib inline
import numpy as np
import pylab

# Data
You are given two datasets consisting of measured rainfall in inches at several stations scattered across North Carolina. There is a column for each station and a row for each day. That is, a 10 by 4 array represents 10 days with data at 4 stations for each day.

In [2]:
site1 = np.array([ [ 35.3,  55. ,  93. ,   9.9],
                   [ 13.9,  34.9,  49.3,  37.7],
                   [ 47.2,   2.2,  15.7,  96.2],
                   [  8.9,  91.7,  12.3,  86. ],
                   [ 32.4,   6.7,  53.1,  93.3]  ])

site2 = np.array([ [ 24. ,  67.9,  16.7,   9.7,  18.8],
                   [ 60.4,  47.3,  44.2,  57. ,  47. ],
                   [ 50.4,  38.6,  60. ,  42.9,  69.2],
                   [ 63.4,  35.5,  10.9,   4. ,  28. ],
                   [  7.4,  42.6,  62.6,  39.1,   0.9],
                   [ 57.6,  13.8,  25.2,  27. ,  34.5]])

## Questions
You must create functions for each question that will analyze site1 and site2. They have a different number of days and stations. Thus, you will not be able to assume the number of stations or days in your code. 

## 1. How many numbers are in the arrays?
Create a funciton called "f1" that will return the amount of numbers in an array. I will show you number 1 as a guide for the other questions.

In [3]:
#You don't have to change anything in this cell. It's just to show you how it's done. 

#Define f1 here 

def f1(dataSet):
    '''Finds the amound of numbers in a data set'''
    return np.product(dataSet.shape)

#I will call the funciton on both sites
print(str(f1(site1)) + "\n" + str(f1(site2)))

20
30


I got 20 and 30 respectively

## 2. What was the total rainfall for each day?
Create a function called "f2" that will return an array representing the total rainfall for each day.

In [4]:
def f2(A):
    return np.sum(A, axis=1)#Define f2 here




#I will call the funciton on both sites
print(str(f2(site1)) + "\n" + str(f2(site2)))

[ 193.2  135.8  161.3  198.9  185.5]
[ 137.1  255.9  261.1  141.8  152.6  158.1]


I got [ 193.2,   135.8,   161.3,   198.9,   185.5] and [ 137.1,  255.9,  261.1,  141.8,  152.6,  158.1] respectively

## 3. Which station had the highest total rainfall?
Create a funciton called "f3" that will return the number of the station with the highest total rainfall.

In [11]:
def f3(A):
    return np.argmax(np.max(A, axis=0))#Define f3 here




#I will call the funciton on both sites
print(str(f3(site1)) + "\n" + str(f3(site2)))

3
4


I got station 3 and station 0 respectively

## 4. Which station had the lowest rainfall on a single day?
Create a funciton called "f4" that will return the number of the station with the lowest rainfall on a single day.

In [6]:
def f4(A):
    return np.argmin(np.min(A, axis=0))#Define f4 here




#I will call the funciton on both sites
print(str(f4(site1)) + "\n" + str(f4(site2)))

1
4


I got station 1 and station 4 respectively

## 5. Which rows have a total rainfall above 190?
Create a function called "f5" that returns a new array. This array will only have data from rows of rain for where the total is above 190. 

In [7]:
def f5(A):
    B= []
    for i in range(len(A[0])):
        if np.sum(A[i]) > 190:
            B.append(A[i])
    return B#Define f5 here




#I will call the funciton on both sites
print(str(f5(site1)) + "\n\n" + str(f5(site2)))

[array([ 35.3,  55. ,  93. ,   9.9]), array([  8.9,  91.7,  12.3,  86. ])]

[array([ 60.4,  47.3,  44.2,  57. ,  47. ]), array([ 50.4,  38.6,  60. ,  42.9,  69.2])]


I got [[ 35.3,  55. ,  93. ,   9.9],  [  8.9,  91.7,  12.3,  86. ]] for site1 and [[ 60.4,  47.3,  44.2,  57.,   47.],[ 50.4,  38.6,  60.,   42.9,  69.2]] for site2

## 6. What are the rainfall amounts for all stations on the day with the lowest total rainfall?
Create a function called "f6" that returns a 1D array with the values for all stations on the day that the minimum occurred.

In [8]:
def f6(A):
    return A[np.argmin(np.min(A, axis=0)),:]#Define f6 here




#I will call the funciton on both sites
print(str(f6(site1)) + "\n" + str(f6(site2)))

[ 13.9  34.9  49.3  37.7]
[  7.4  42.6  62.6  39.1   0.9]


I got [ 13.9,  34.9,  49.3,  37.7] and [ 24.,   67.9,  16.7,   9.7,  18.8] respectively

## 7. What was the total rainfall on odd numbered days?
Create a funciton called "f7" that returns the total rainfall of the odd numbered days.

In [9]:
def f7(A):
    B= 0
    for i in range(len(A[0])):
        if i%2 != 0:
            B= B + np.sum(A[i])
    return B#Define f7 here




#I will call the funciton on both sites
print(str(f7(site1)) + "\n" + str(f7(site2)))

334.7
397.7


I got 334.7 and 555.8 respectively

## Done!

In [19]:
import comp116
comp116.submit('ws11')